src/librustc_traits/lowering.rs

```diff
@@ -260,6 +276,39 @@ fn implied_bound_from_trait<'a, 'tcx>(
     }))
 }
 
+fn wellformed_from_bound<'a, 'tcx>(
+    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    trait_pred: ty::TraitPredicate<'tcx>,
+    where_clause: &ty::Predicate<'tcx>,
+) -> Clause<'tcx> {
+    // Rule WellFormed-TraitRef
+    //
+    // For each where clause WC:
+    // forall<Self, P1..Pn> {
+    //   WellFormed(Self: Trait<P1..Pn>) :- Implemented(Self: Trait<P1..Pn>) && WellFormed(WC)
+    // }
```

Comment:
Actually the guide is a bit misleading here, `WC` should represent the full set of where clauses e.g. you have `WC := WC1, WC2, ..., WCn` and we want to output only one clause which is:

```
WellFormed(Self: Trait<P1..Pn>) :-
    Implemented(Self: Trait<P1..Pn>) && WC1 && ... && WCn
```

Reply:
(So you should change your code logic a bit)

Reply:
```
Implemented(Self: Trait<P1..Pn>) && WC1 && ... && WCn
                                    ^^^ = should this be WellFormed(WC1)?
```

Since the wellformed [guide](https://rust-lang-nursery.github.io/rustc-guide/traits-lowering-rules.html#implied-bounds) says:
>  T: Trait is well-formed if (a) T: Trait is implemented and (b) all the where-clauses declared on Trait are well-formed

Thus do you mean this(merge all the `WellFormed(WC_1...n)` inside one clause instead of split them like current test/ui output below)?


Currently(copy from ui test output)
```
   = note: WellFormed(Self: Foo<'a, 'b, S, T, U>) :- Implemented(Self: Foo<'a, 'b, S, T, U>), RegionOutlives('a : 'b).
   = note: WellFormed(Self: Foo<'a, 'b, S, T, U>) :- Implemented(Self: Foo<'a, 'b, S, T, U>), TypeOutlives(U : 'b).
   = note: WellFormed(Self: Foo<'a, 'b, S, T, U>) :- Implemented(Self: Foo<'a, 'b, S, T, U>), WellFormed(S: std::fmt::Debug).
   = note: WellFormed(Self: Foo<'a, 'b, S, T, U>) :- Implemented(Self: Foo<'a, 'b, S, T, U>), WellFormed(S: std::marker::Sized).
   = note: WellFormed(Self: Foo<'a, 'b, S, T, U>) :- Implemented(Self: Foo<'a, 'b, S, T, U>), WellFormed(T: std::borrow::Borrow<U>).
   = note: WellFormed(Self: Foo<'a, 'b, S, T, U>) :- Implemented(Self: Foo<'a, 'b, S, T, U>), WellFormed(T: std::marker::Sized).
```
After=>
```
= note: WellFormed(Self: Foo<'a, 'b, S, T, U>) :- Implemented(...), ..., WellFormed(WC1) , ..., WellFormed(WCn).
```

Reply:
Yes absolutely, merge all into one clause (and yes you’re right about `WellFormed(WC1), ..., WellFormed(WCn)`